### ISSUE TITLE
CONFIGpac checksums are always wrong

### ISSUE THREAD

**Issue by 314gomes** [reedhedges/AriaCoda#15]
Whenever I start a new connection with my P3DX-SH, I receive get the following warning:
`ArRobotPacketReceiver::receivePacket: Warning: bad packet, bad checksum (received packet type ID 0x20: Robot Config Packet)`

I dug deeper into AriaCoda's code and found out that there really is always a mismatch between CONFIGpac's reported checksum and calculated checksum, but it seems to me that according to the robot's manual (which I could only find here https://www.inf.ufrgs.br/~prestes/Courses/Robotics/manual_pioneer.pdf ) robot configuration packets seem to have no trailing checksum.

Would it make sense to ignore the packet checksum if it is a CONFIGpac? I could open a PR for that.

**Comment by 314gomes** [comment]
By the way, here is the data from one of the CONFIGpacs
[0] 0xfa 
[1] 0xfb 
[2] 0x89 
[3] 0x20 
[4] 0x50 
[5] 0x69 
[6] 0x6f 
[7] 0x6e 
[8] 0x65 
[9] 0x65 
[10] 0x72 
[11] 0x0 
[12] 0x70 
[13] 0x33 
[14] 0x64 
[15] 0x78 
[16] 0x2d 
[17] 0x73 
[18] 0x68 
[19] 0x0 
[20] 0x47 
[21] 0x4a 
[22] 0x43 
[23] 0x5f 
[24] 0x33 
[25] 0x31 
[26] 0x35 
[27] 0x36 
[28] 0x0 
[29] 0x0 
[30] 0x68 
[31] 0x1 
[32] 0xe8 
[33] 0x3 
[34] 0x2c 
[35] 0x1 
[36] 0xd0 
[37] 0x7 
[38] 0xf4 
[39] 0x1 
[40] 0x46 
[41] 0x41 
[42] 0x50 
[43] 0x45 
[44] 0x53 
[45] 0x50 
[46] 0x5f 
[47] 0x33 
[48] 0x31 
[49] 0x35 
[50] 0x36 
[51] 0x0 
[52] 0x64 
[53] 0x0 
[54] 0x0 
[55] 0x0 
[56] 0x0 
[57] 0x1 
[58] 0x0 
[59] 0x1 
[60] 0x73 
[61] 0x0 
[62] 0xba 
[63] 0x40 
[64] 0xd0 
[65] 0x7 
[66] 0x0 
[67] 0xc8 
[68] 0x0 
[69] 0xc8 
[70] 0x0 
[71] 0xe8 
[72] 0x3 
[73] 0x32 
[74] 0x0 
[75] 0x64 
[76] 0x0 
[77] 0xee 
[78] 0x2 
[79] 0x64 
[80] 0x0 
[81] 0x64 
[82] 0x0 
[83] 0x28 
[84] 0x0 
[85] 0x14 
[86] 0x0 
[87] 0x0 
[88] 0x0 
[89] 0x2c 
[90] 0x1 
[91] 0x2c 
[92] 0x1 
[93] 0x28 
[94] 0x0 
[95] 0x1e 
[96] 0x0 
[97] 0x0 
[98] 0x0 
[99] 0x0 
[100] 0x0 
[101] 0x0 
[102] 0x28 
[103] 0x2 
[104] 0x0 
[105] 0x0 
[106] 0x0 
[107] 0x0 
[108] 0x0 
[109] 0x80 
[110] 0x0 
[111] 0x6c 
[112] 0x0 
[113] 0x32 
[114] 0x2e 
[115] 0x33 
[116] 0x0 
[117] 0xfd 
[118] 0x19 
[119] 0x85 
[120] 0x6 
[121] 0x0 
[122] 0x0 
[123] 0x0 
[124] 0x0 
[125] 0x0 
[126] 0x0 
[127] 0x0 
[128] 0x0 
[129] 0x0 
[130] 0x0 
[131] 0x0 
[132] 0x0 
[133] 0x0 
[134] 0xff 
[135] 0x4 
[136] 0x11 
[137] 0xff 
[138] 0xa1 
[139] 0x32

**Comment by reedhedges** [comment]
What happens if you connect to your robot using the "demo" example program, and press 'o' to request a CONFIG packet after the connection has been made?  Do you still get a bad checksum?  (Also, do you get this warning at any other time?)   If the checksum is always wrong then that's a bug in the P3DX firmware (ARCOS) but other robot types have different firmware that may provide a correct checksum, so ignoring the checksum for any CONFIG packet would, in my opinion, not be a good idea... Unless this problem impacts the ability to use the robot then we would need to come up with some solution:  I think you should still be able to read the information with ArConfigPacketReader, is this the case?    Unfortunately I don't have any robots to test with anymore (just the simulator).

**Comment by 314gomes** [comment]
Upon pressing 'o' in ariaDemo, I get the following output
```
ArRobotPacketReceiver::receivePacket: Warning: bad packet, bad checksum (received packet type ID 0x20: Robot Config Packet)
ArRobot::myPacketReader: Timed out (4) at 28505 (200 into cycle after sleeping 135)
```

The received robot config packet is coming with bad checksums every time command 18 is sent too.

I am not able to read any information using robot->getOrigRobotConfig(). The returned ArRobotConfigPacketReader has default values in every parameter. Is there any other way of using ArRobotConfigPacketReader?

As to the firmware, could this be the result of a flashing error? I was unable to find any download to the tool that used to be available for reflashing ARCOS. Has it been archived anywhere?

**Comment by 314gomes** [comment]
Where could I find the ARCOScf tool? http://robots.mobilerobots.com has been down for a very long time and the programs have not been archived by archive.org

I did have a look inside AriaCoda and found nothing except that it seems like CONFIGpac's weren't supposed to have a checksum in the first place (at least according to the manual PDF), but it seems weird to me that something like that would've gone unnoticed for so many years. Do you know of any other robot with this problem?

**Comment by reedhedges** [comment]
I would expect CONFIG packets to always have checksums, the same as any other packet.  As far as I know this is the case in MobileSim/AMRISim, and in all firmware variants (AROS, P2OS, SeekurOS, Lynx/PioneerLX, etc.) including ARCOS but I don't have any way to check or verify any of those except AMRISim (Which seems to send a normal packet with checksum.)    You can use command line arguments to enable more packet logging (`-robotLogPacketsReceived`), and we can add more to `ArRobotPacketReceiver` if that is helpful.  (In fact you can see some extra debugging logging commented out there that you could include and recompile if desired.) or attach your own callback with `ArRobotPacketReceiver::setPacketReceivedCallback()`.

**Comment by reedhedges** [comment]
(Note There's also a global debugging flag for all "device connections" (serial port or TCP port connections for the robot and accessory hardware devices) that can be set via `ArDeviceConnection::debugShouldLog(true);`, this is a static method of ArDeviceConnection so it affects all instances.)

**Comment by 314gomes** [comment]
It seems like the last two bytes of the CONFIG packets I'm receiving are actually some sort of checksum: I uncommented some code to print the packet content when the calculated checksum didn't match the expected and indeed the last two bytes did change after changes in other parts of CONFIG packets.

I am led to believe that a mistake happened when updating the ArRobotPacket::calcCheckSum() function, as when I changed it back to what appears in the archived manual (https://web.archive.org/web/20180613215151/http://robots.mobilerobots.com:80/docs/all_docs/P3OpMan6_5.pdf page 24) suddenly I was not having errors on config packets anymore. I believe this section of code was changed on a very big commit 7046fd2 substituting the use of Aria-defined types to standard cpp type.

**Comment by 314gomes** [comment]
I found something meaningful. In this line
https://github.com/reedhedges/AriaCoda/blob/6afa999907519c7ff9cebadba3dfced830f0c728/src/ArRobotPacket.cpp#L122
the casting from myBuf[] (char type) to int is sometimes returning a negative value, which breaks the checksum calculation. Replacing this line with
```cpp
  unsigned char n = myBuf[2] - 2;
```
makes the returned checksum match the expected.

**Comment by 314gomes** [comment]
If you want, I can create a PR with this simple change.

**Comment by reedhedges** [comment]
Yes please do, thanks! If anyone has any ideas on creating some good tests
for checksum calculations let us know and we can try to add them.

**Comment by 314gomes** [comment]
I opened the PR. Please do tell me if I did anything that would not be recommended as this is my first commit in an open source project.

**Comment by reedhedges** [comment]
Thanks for the contribution, I merged it though may add a comment explaining why the cast is ok etc.   If you are sure that the main problem (failed checksums on packets) seems to be fixed, then we can consider this bug closed.  If anyone else is able to test this, that would be appreciated.   If you (João) happen to have still have any example packet data that triggered the bug (in addition to what you sent earlier in this discussion), maybe we can set up a test case to specifically test it?  If you or anyone else is interested in setting up that kind of test case, that would be great, ARIA never had a good set of automated tests though I did try to start preparing for better automated testing in AriaCoda a while back.

**Comment by 314gomes** [comment]
Yes, I think we could consider this bug fixed as I'm now able to correctly read my robot's initial configuration. I could supply the project with new test cases for packet error checking, but I would need a bit more help as I did not yet correctly understand how the current test suite works. Maybe open a new issue for that?
